k8srm-prototype/pod-spec-examples.yaml

```diff
@@ -0,0 +1,133 @@
+# could eliminate that with equivalent expressiveness, by instead listing
+# "allOf" and "oneOf", where "oneOf" is actually a list-of-lists of claims, but
+# not sure that's better.
+
+
+# Allocates four SR-IOV VFs from any Intel NIC These could even each come from
+# a different NIC (on the same node of course). NICs are not tied to specific
+# containers, so there is no need to name the claim; the NICs will be added to
+# the pod-level sandbox.
+deviceClaims:
+- template:
+    claims:
+    - deviceClass: "intel-sriov-vf"
+      requests:
+        count: 4
+
+# Allocates four SR-IOV VFs from any vendor NIC, making sure all the VFs are on
+# the same PF this requires vendor & PF together to fully qualify a physical
+# card on a node.
+deviceClaims:
+- template:
+    claims:
+    - deviceType: "sriov-vf"
+      count: 4
+      matchAttributes:
+      - "vendor"
+      - "pf"
+
+# Allocates two A100 GPUs on the same PCI root complex. The name is used to
+# attach the associated claims to a container.
+deviceClaims:
+- template:
+    claims:
+    - name: gpus
+      deviceClass: "nvidia-a100"
+      count: 2
+      matchAttributes:
+      - pcie-root
+
+# Allocates two A100 GPUs on the same PCI root complex, but keeps them
+# separated in the API so they can be attached to two different containers.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu-one
+      deviceClass: "nvidia-a100"
+    - name: gpu-two
+      deviceClass: "nvidia-a100"
+
+# Allocates an SR-IOV PF and a an A100 from the same pcie-root. Name is
+# populated here for the PF just to show that it can be, but the expectation is
+# generally for network devices it is not needed. Note ANY vendor PF is fine
+# here; we use deviceType, not deviceClass.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu
+      deviceClass: "nvidia-a100"
+    - name: nic
+      deviceType: "sriov-pf"
+
+# Allocates either an A100, H100, or two L4s that are on the same pcie root.
+deviceClaims:
+- template:
+    claims:
+    - name: gpus
+      oneOf:
+      - deviceClass: "nvidia-a100"
+      - deviceClass: "nvidia-h100"
+      - deviceClass: "nvidia-l4"
+        count: 2
+        matchAttributes:
+        - pcie-root
+
+# Allocates and SR-IOV VF from any vendor and either an A100, H100, or two L4s,
+# all from the same NUMA node and PCIe root complex.
+deviceClaims:
+- template:
+    matchAttributes:
+    - numa
+    - pcie-root
+    claims:
+    - name: nic
+      deviceType: "sriov-vf"
+    - name: gpus
+      oneOf:
+      - deviceClass: "nvidia-a100"
+      - deviceClass: "nvidia-h100"
+      - deviceClass: "nvidia-l4"
+        count: 2
+
+# Allocates 4 VFs from 4 different PFs,
+# but all the same vendor
+deviceClaims:
+- template:
+  claims:
+  - name: nics
+    deviceType: "sriov-vf"
+    count: 4
+    matchAttributes:
+    - vendor
+    distinctAttributes: # maybe we should use affinity/antiaffinity terms?
```

Comment:
this feels like YAGNI - do we have concrete needs for this?  Another term we use in some places is "spread".  Or "unionAttribs" and "disjointAttribs" or "disjuntionAttribs" ?

Reply:
it could create 4 claims in this case, one per PF, no?

Reply:
You would have to know about the infrastructure ahead of time to do that. What if there are 16 PFs? We just want to be on different ones,  not specific ones.

Reply:
for networking use cases you'll always have to be aware, at least is what I learnt from the Telco use cases, those nics will be attached to "something" externally that was already preprovisioned ...

Reply:
If they are pre-provisioned, then yes, you could explicitly select individual ones. My thinking for this though is more like a redundancy use case, and they need not be pre-provisioned in that case. Instead, I could attach a config object that specifies which external network to attach to. The driver would then consume that object and plumb the appropriate networking on that node. This decouples the selection (I want 4 VFs), the constraints on that selection (VFs on different PFs), and the configuration (configure those VFs all to attach to VLAN 2000).

So, imagine deploying some network function that needs to attach to the "internet" external network, and you want it to use two VFs. You can increase availability if those VFs are on different PFs. So, you create a claim that references a CR that says "attach to the 'internet' external network". The claim also specifies two VFs, with "distinct" PFs. The scheduler allocates those. What reaches the driver is "add these devices (pf0vf0 and pf1vf0) to the pod" and "configure the devices access to the 'internet' external network". The driver then needs to interpret those two things and "make it so".